accord-core/src/main/java/accord/local/Command.java

```diff
@@ -19,362 +19,487 @@
+        if (executeAt != null && hasBeen(Committed) && !this.executeAt().equals(executeAt))
+            safeStore.agent().onInconsistentTimestamp(this, this.executeAt(), executeAt);
+        setDurability(durability);
     }
 
-    // requires that command != null
-    // relies on mutual exclusion for each key
-    // note: we do not set status = newStatus, we only use it to decide how we register with the retryLog
-    private void witness(Txn txn, Key homeKey, Key progressKey)
+    public enum AcceptOutcome
     {
-        txn(txn);
-        homeKey(homeKey);
-        progressKey(progressKey);
-
-        if (status() == NotWitnessed)
-            status(PreAccepted);
+        Success, Redundant, RejectedBallot
+    }
 
-        if (executeAt() == null)
-        {
-            Timestamp max = commandStore().maxConflict(txn.keys());
-            // unlike in the Accord paper, we partition shards within a node, so that to ensure a total order we must either:
-            //  - use a global logical clock to issue new timestamps; or
-            //  - assign each shard _and_ process a unique id, and use both as components of the timestamp
-            Timestamp witnessed = txnId().compareTo(max) > 0 && txnId().epoch >= commandStore().latestEpoch()
-                    ? txnId() : commandStore().uniqueNow(max);
-            executeAt(witnessed);
+    public AcceptOutcome preaccept(SafeCommandStore safeStore, PartialTxn partialTxn, AbstractRoute route, @Nullable RoutingKey progressKey)
+    {
+        if (promised().compareTo(Ballot.ZERO) > 0)
+            return AcceptOutcome.RejectedBallot;
 
-            txn.keys().foldl(commandStore().ranges().since(txnId().epoch), (i, key, param) -> {
-                if (commandStore().hashIntersects(key))
-                    commandStore().commandsForKey(key).register(this);
-                return null;
-            }, null);
-        }
+        return preacceptInternal(safeStore, partialTxn, route, progressKey);
     }
 
-    public boolean preaccept(Txn txn, Key homeKey, Key progressKey)
+    private AcceptOutcome preacceptInternal(SafeCommandStore safeStore, PartialTxn partialTxn, AbstractRoute route, @Nullable RoutingKey progressKey)
     {
-        if (promised().compareTo(Ballot.ZERO) > 0)
+        if (executeAt() != null)
         {
-            logger.trace("{}: skipping preaccept - witnessed higher ballot ({})", txnId(), promised());
-            return false;
+            logger.trace("{}: skipping preaccept - already preaccepted ({})", txnId(), status());
+            return AcceptOutcome.Redundant;
         }
 
-        if (hasBeen(PreAccepted))
+        Status status = status();
+        switch (status)
         {
-            logger.trace("{}: skipping preaccept - already preaccepted ({})", txnId(), status());
-            return true;
+            default: throw new IllegalStateException();
+            case NotWitnessed:
+            case AcceptedInvalidate:
+            case Invalidated:
```

Comment:
Why do we fall through for these three? If we invalidated or accepted invalidate we can still pre-accept?

Reply:
In the case of `Invalidated` it's at least redundant to `PreAccept`, so I've updated to reflect that. However, yes, we can definitely pre-accept if we have `AcceptedInvalidate` - as we may yet move to `Accepted`; `AcceptedInvalidate` is not a terminal state.